Use hie-bios to generalize build command

alt-romes opened the pull request:
Currently, HDA determines how to run ghci for any given project by getting runCmd from the client.

Even though the VSCode extension and others set a default scaffolding for either stack or cabal projects, this approach has a few shortcomings:

1. Newish Haskell users are not well versed in the build systems so they can't work around failures to invoke the commands as specified

2. Other build systems or even simple Haskell programs that are not a package at all cannot benefit from HDA.

    - For example, GHC uses the hadrian build system.
    - Users new to Haskell don't use cabal nor stack straight away, but could benefit from step by step execution of their program

Luckily, the hie-bios project aims to solve exactly this problem of determining how to invoke GHC for any given Haskell project and supports a wide range of different tools. It also allows users to be very precise about their own build set up by reading a hie.yaml configuration.

haskell-language-server already uses this approach: call functions from hie-bios to determine how to invoke ghc.

This commit does the same for HDA, to make it more robust across different Haskell projects.

alt-romes committed:
Use hie-bios to generalize build command

Currently, HDA determines how to run ghci for any given project by
getting runCmd from the client.

Even though the VSCode extension and others set a default scaffolding
for either stack or cabal projects, this approach has a few
shortcomings:

1. Newish Haskell users are not well versed in the build systems so they
   can't work around failures to invoke the commands as specified

2. Other build systems or even simple Haskell programs that are not a
   package at all cannot benefit from HDA.

    - For example, GHC uses the hadrian build system.
    - Users new to Haskell don't use cabal nor stack straight away, but
      could benefit from step by step execution of their program

Luckily, the hie-bios project aims to solve exactly this problem of
determining how to invoke GHC for any given Haskell project and supports
a wide range of different tools. It also allows users to be very precise
about their own build set up by reading a hie.yaml configuration.

haskell-language-server already uses this approach: call functions from
hie-bios to determine how to invoke ghc.

This commit does the same for HDA, to make it more robust across
different Haskell projects.

Manually tested on:
- GHC (custom hie.yaml)
- fast-tags (Cabal package)
- Simple Main.hs file with 1 dependency

alt-romes commented:
I have tested debugging the following different types of projects using this patch successfully:

- GHC itself (custom hie.yaml)
- Simple Main.hs file with 1 dependency (no project)
- fast-tags (a Cabal package)

On the other hand, Stack is problematic. I don't have a lot of experience with it, however, from what I understand, a stack project can specify the GHC version used and stack will accordingly select it.

The problem is that, unlike `haskell-language-server` (which has a wrapper which chooses the appropriate GHC version), `haskell-debug-adapter` always calls the `ghci-dap` which is fixed to a specific version. If the version of the ghci-dap doesn't match the version of GHC used by `stack` (which builds the dependencies with that version and writes them to a package database that the repl depends on), it ultimately crashes when the stack project is loaded in it.

How does `haskell-debug-adapter` currently solve this? Is it that passing `--with-ghc` to stack repl will override the resolver version?

If not, perhaps we could merge this first and think about how to handle a wide range of versions in a uniform way (à la HLS) in subsequent updates.

@phoityne perhaps you could try this patch out? how do you usually test these changes?

phoityne commented:
Thank you for the great PR. 
I've tried it in local environment.
It worked well with single source file.
But could not break on linked other source code.
It does not import Lib.hs file.

(hie-bios)
```
Now, waiting for an initial prompt("ghci> ") from ghci.

[DAP][INFO] start ghci-dap-0.0.25.0.
GHCi, version 9.12.1: https://www.haskell.org/ghc/  :? for help
[1 of 2] Compiling Main             ( app/Main.hs, interpreted )
Ok, one module loaded.
ghci> 
ghci> [DEBUG][APP] GHCi version is 9.12.1
:set prompt "<<DAP_CMD_END2>>\nH>>= "
H>>= :set prompt-cont "<<DAP_CMD_END2>>\nH>>= "
H>>=
H>>= :dap-launch ...
H>>= :dap-context-modules ...
[DAP][INFO]context modules. ["Main"]
H>>= :dap-set-breakpoints ...
```

(cabal)
```
Now, waiting for an initial prompt("ghci> ") from ghci.

Configuration is affected by the following files:
- cabal.project.local
[DAP][INFO] start ghci-dap-0.0.25.0.
Resolving dependencies...
Build profile: -w ghc-9.12.1 -O1
In order, the following will be built (use -v for more details):
 - sample-cabal-0.1.0.0 (interactive) (lib) (first run)
Preprocessing library for sample-cabal-0.1.0.0...
[DAP][INFO] start ghci-dap-0.0.25.0.
GHCi, version 9.12.1: https://www.haskell.org/ghc/  :? for help
ghci> [DEBUG][APP] GHCi version is 9.12.1
:set prompt "<<DAP_CMD_END2>>\nH>>= "
H>>= :set prompt-cont "<<DAP_CMD_END2>>\nH>>= "
H>>= :dap-launch ...
H>>= :load /work/haskell/sample-cabal/app/Main.hs
[1 of 3] Compiling Lib              ( src/Lib.hs, interpreted )
[2 of 3] Compiling Main             ( /work/haskell/sample-cabal/app/Main.hs, interpreted )
Ok, two modules loaded.
H>>=
H>>= :dap-context-modules ...
[DAP][INFO]context modules. ["Lib","Main"]
H>>= :dap-set-breakpoints ...
```

Is this means debugger(with hie-bios) for single file ? 
> Simple Main.hs file with 1 dependency (no project)

alt-romes commented:
Hi @phoityne 

I mean that simple files that import other files should all work. So,
```hs
module Main where
import Hi

main = do
  putStrLn "main"
  hi
```

```hs
module Hi where
hi = putStrLn "hi"
```

both of these modules are loaded, and I can set the breakpoint in VSCode in `Hi.hs` without problems.

> [1 of 2] Compiling Main             ( app/Main.hs, interpreted )
> Ok, one module loaded.
> ghci> 

In your example, it looks like `Main` does not depend on `Lib` from this error message.

With the current design, for simple projects with just a few modules, only the modules that Main depends on are loaded.
This is because it is GHC's --make mode who realizes what modules must be loaded in this case.

alt-romes commented:
(Fixes #36)

phoityne commented:
Hi.
I confirmed that it works with multiple source files.
I wonder if we could leave the current implementations for cabal and stack as is, and add your implementation that uses hie-bios as a separate function.

Then, add a launch configuration for using hie-bios.
```
launch.json
{
"type": "ghc",
"request": "launch",
"name": "haskell", # or such as "haskell(hie-bios)"
"ghciCmd": "ghci-dap",
}
```
If the ghciCmd string is "ghci-dap", run your function that uses hie-bios.

Regards.

alt-romes commented:
> I wonder if we could leave the current implementations for cabal and stack as is, and add your implementation that uses hie-bios as a separate function.

This is a good idea, I'm going to update the PR.

alt-romes committed:
Only enable hie-bios when cmd=ghci-dap

Re-introduce stack and cabal based debugging sessions.
Now, hie-bios will only be used if the ghciCmd in launch.json is exactly
"ghci-dap"

This allows hie-bios workflows to co-exist with the remaining ones,
serving as a safety net if something goes wrong.

alt-romes commented:
I've updated the commit to support all three configurations.
I've also pushed PRs to hdx4vim and hdx4vsc to use the new configuration

phoityne commented:
Thank you for your contribution.
It works well.
But can not build with ghc-8.10.7, because version restriction hie-bios>=0.14.
Could you remove the restriction ?
I'll check hdx4vim and hdx4vsc later.

alt-romes commented:
I've updated the patch to work with 8.10 as well now. Ready !

phoityne commented:
Thank you!
I have successfully built it with GHC 8.10.7 and merged the PR.
